### PR TITLE
added ability for plugins to provide additional params to pass to proot

### PR DESCRIPTION
**Pull request by kadrim**
addresses issue #48

### CODE REVIEW

**Review by xeffyr — CHANGES_REQUESTED** [review]
If you want to implement QEMU support, follow the general style of `proot-distro` script. Do not create unnecessary plugins and extend functionality through modifying `proot-distro.sh`.

For example you can introduce an optional variable `DISTRO_ARCH` which will allow to override detected architecture and run distribution through QEMU.

File: proot-distro.sh [hunk 1]
```diff
@@ -1089,6 +1105,24 @@ command_login() {
 			set -- "-p" "$@"
 		fi
 
+		# Load Plugin-Script for special handling if needed
```
[the remark this answers is not in the document]
Comment (kadrim) [reply proot-distro.sh:1108]: do you have a suggestion how i can detect what params should be appended to proot without loading the plugin script?

i could write another file to /etc/ during setup which will save the distros architecture and modify the script here to check if distro-arch != host-arch and then append a corresponding param.

but ist that really endorsed?

**Comment by kadrim** [comment]
> For example you can introduce an optional variable `DISTRO_ARCH` which will allow to override detected architecture and run distribution through QEMU.

good idea - can i introduce a new CLI argument, so that the user can decide which arch he wants to use (override)?

File: proot-distro.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (xeffyr) [reply proot-distro.sh:1108]: Add comparisons like this one into `proot-distro.sh`:
``` .bash
...
if [ "$DISTRO_ARCH" != "$(uname -m)" ] && [ "$DISTRO_ARCH" = "i386" ]; then
   set -- "-q" "$PREFIX/bin/qemu-i386"
fi
...
```
Also define `DISTRO_ARCH` to device arch (`uname -m`) somewhere before plugins load. In plugins, replace `$(uname -m)` calls with `$DISTRO_ARCH`.

That will allow to override the device arch and tell proot to use QEMU.

**Comment by xeffyr** [comment]
> good idea - can i introduce a new CLI argument

Not a cli argument. Config variable in distro plugin.

I perhaps will have to do that on my own.

**Comment by kadrim** [comment]
@xeffyr good, that you made those adjustments so quickly!

Referring to your comment

> Using a separate plugin for SAME distribution but with QEMU support discards their purpose. You may create only one plugin per distribution version.

 with the solution you implemented, one has to add another (or modify the) plugin to get it to work with qemu.

i think, that using a cmdline arg like `--force-arch=...` would be better.

I can open an issue and provide the pull-request for that if you like.

**Comment by xeffyr** [comment]
> with the solution you implemented, one has to add another (or modify the) plugin to get it to work with qemu.

Yes

> i think, that using a cmdline arg like --force-arch=... would be better.

And user will have to specify a right architecture during installation, then specify it each time `proot-distro login` was invoked, etc. 

In order to add flag `--force-arch`, I will have to rewrite completely the "plugin system" in order to track user's preferences specified during installation time because now it doesn't have a way to do that. That's something that will not be done, at least not now. QEMU support is not worth releasing a new backwards-incompatible major version.

`proot-distro` was designed to offload as much as possible proot configuration to automatics and considering this, runtime option `--force-arch` is really something redundant/unwanted.

QEMU user mode is not a part of primary `proot-distro` functionality. It is also not stable, despite it is in the main (stable) repository.

**Comment by kadrim** [comment]
it totally understand your concerns and even more your reasons.

with only slight adjustments, i could make it backwards compatible with your current "plugin system".

for example: using the `--force-arch` option will always create an alias so an .override file in /etc/ which will set the `DISTRO_ARCH` variable.

I know you already invested too much time in a feature you never wanted, but maybe you will let me help you out with this one.

Thanks in advance and best regards!

**Pull request closed without merging**